lib/gps.js

```diff
@@ -0,0 +1,367 @@
+  Object.defineProperties(this, module);
+
+  // Controller decorates the instance
+  if (this.opts.controller) {
+    controller = typeof this.opts.controller === "string" ?
+      Controllers[this.opts.controller] : this.opts.controller;
+
+    Board.Controller.call(this, controller, opts);
+  }
+
+  // If necessary set default property values
+  this.precision = opts.precision || 9;
+  this.baud = opts.baud || controller.baud || device.baud;
+
+  // Create a "state" entry for privately
+  // storing the state of the instance
+  state = {
+    sat: {},
+    pos: {},
+    frequency: 1000,
+    lowPowerMode: false
+  };
+
+  priv.set(this, state);
+
+  // Getters for private state values
+  Object.defineProperties(this, {
+    pos: {
+      get: function() {
+        return state.pos;
+      }
+    }
+  });
+
+  if (this.initialize) {
+    this.initialize(opts);
+  }
+
+}
+
+util.inherits(GPS, events.EventEmitter);
+
+/*
+ * Default intialization for serial GPS
+ */
+GPS.prototype.initialize = function(opts) {
+
+  this.serialPort = opts.serialPort || opts.portId || opts.port || opts.bus || this.io.SERIAL_PORT_IDs.SW_SERIAL0;
+
+  // Set the pin modes
+  ["tx", "rx"].forEach(function(pin) {
+    if (this.pins[pin]) {
+      this.io.pinMode(this.pins[pin], this.io.MODES.SERIAL);
+    }
+  }, this);
+
+  if (this.pins.onOff) {
+    this.io.pinMode(this.pins.onOff, this.io.MODES.OUTPUT);
+    this.onOff = new Pin(this.pins.onOff);
+  }
+
+  this.io.serialConfig({
+    portId: this.serialPort,
+    baud: this.baud,
+    rxPin: this.pins.rx,
+    txPin: this.pins.tx
+  });
+
+  this.configure(function() {
+    this.listen();
+    if (opts.freq) {
+      this.frequency(opts.freq);
+    }
+  }.bind(this));
+
+};
+
+GPS.prototype.sendCommand = function( inString ) {
+  var cc = [];
+
+  // Convert the string to a charCode array
+  for (var i = 0; i < inString.length; ++i) {
+    cc[i] = inString.charCodeAt(i);
+  }
+
+  // Append *, checksum and cr/lf
+  var hexsum = getNmeaChecksum(inString.substring(1));
+  cc.push(42, hexsum.charCodeAt(0), hexsum.charCodeAt(1), 13, 10);
+
+  this.io.serialWrite(this.serialPort, cc);
+};
+
+GPS.prototype.listen = function() {
+  var input = "";
+  // Start the read loop
+  this.io.serialRead(this.serialPort, 0, function(data) {
+
+    input += new Buffer(data).toString("ascii");
+    var sentences = input.split("\r\n");
+
+    if (sentences.length > 1) {
+      for (var i=0; i < sentences.length - 1; i++) {
+        this.parseNmeaSentence(sentences[i]);
+      }
+      input = sentences[sentences.length - 1];
+    }
+  }.bind(this));
+};
+
+/*
+ * NMEA Sentence Information
+ * http://aprs.gids.nl/nmea
+ */
+GPS.prototype.parseNmeaSentence = function(sentence) {
+
+  var state = priv.get(this);
+  var cksum = sentence.split("*");
+
+  // Check for valid sentence
+  if (cksum[1] !== getNmeaChecksum(cksum[0].substring(1))) {
+    return;
+  }
+
+  process.nextTick(this.emit.bind(this, "message", sentence));
```

Comment:
`parseNmeaSentence` is already called asynchronously, because `serialRead` is async. These events don't need to be explicitly scheduled for the next turn